core/src/main/java/com/linecorp/armeria/internal/common/stream/PrependingPublisher.java

```diff
@@ -0,0 +1,135 @@
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public final class PrependingPublisher<T> implements Publisher<T> {
+
+    private final T first;
+    private final Publisher<? extends T> rest;
+
+    public PrependingPublisher(T first, Publisher<? extends T> rest) {
+        this.first = first;
+        this.rest = rest;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        final RestSubscriber restSubscriber = new RestSubscriber(subscriber);
+        rest.subscribe(restSubscriber);
+    }
+
+    final class RestSubscriber implements Subscriber<T>, Subscription {
+
+        private Subscriber<? super T> downstream;
+        private final AtomicLong demand = new AtomicLong();
+        @Nullable
+        private volatile Subscription upstream;
+        @Nullable
+        private volatile Throwable upstreamCause;
+        private volatile boolean upstreamCompleted;
+        private volatile boolean completed;
+        private volatile boolean firstSent;
+
+        RestSubscriber(Subscriber<? super T> downstream) {
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            this.upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T t) {
+            demand.decrementAndGet();
+            downstream.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            // delay onError until the first piece is sent
+            if (!firstSent) {
+                upstreamCause = t;
+            } else {
+                downstream.onError(t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            // delay onComplete until the first piece is sent
+            if (!firstSent) {
+                upstreamCompleted = true;
+            } else {
+                downstream.onComplete();
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (completed) {
+                return;
+            }
+            if (demand.getAndAdd(n) > 0) {
+                return;
+            }
+            if (!firstSent) {
+                firstSent = true;
+                downstream.onNext(first);
+                if (n < Long.MAX_VALUE) {
+                    demand.decrementAndGet();
+                }
+            }
+            if (demand.get() > 0) {
+                if (upstreamCause != null) {
+                    downstream.onError(upstreamCause);
```

Comment:
How about introducing a local variable for `upstreamCause` to avoid accessing the volatile variable twice?